CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/StoreGraph.java

```diff
@@ -2223,43 +2200,13 @@ private void updateLayerMask(final int attributeId, final GraphElementType eleme
 
     // update the layer visibility for an element when an attribute changes
     private void updateLayerMask(final int attributeId, final int elementid) {
-        if (transactionLayerVisibilityAttributeId == Graph.NOT_FOUND) {
-            transactionLayerVisibilityAttributeId = this.addAttribute(
-                    GraphElementType.TRANSACTION,
-                    FloatAttributeDescription.ATTRIBUTE_NAME,
-                    LAYER_VISIBILITY_ATTRIBUTE_LABEL,
-                    "The visibility of the transaction given "
-                    + "the layers it belongs to", 1.0f, null);
-        }
-        if (vertexLayerVisibilityAttributeId == Graph.NOT_FOUND) {
-            vertexLayerVisibilityAttributeId = this.addAttribute(
-                    GraphElementType.VERTEX,
-                    FloatAttributeDescription.ATTRIBUTE_NAME,
-                    LAYER_VISIBILITY_ATTRIBUTE_LABEL,
-                    "The visibility of the vertex given "
-                    + "the layers it belongs to", 1.0f, null);
-        }
-        if (layerMaskSelectedAttributeId == Graph.NOT_FOUND) {
-            layerMaskSelectedAttributeId = this.addAttribute(
-                    GraphElementType.GRAPH,
-                    IntegerAttributeDescription.ATTRIBUTE_NAME,
-                    LAYER_MASK_SELECTED_ATTRIBUTE_LABEL,
-                    "The layers currently enabled for display", 1, null);
-        }
-        if (vertexLayerMaskAttribureId == Graph.NOT_FOUND) {
-            vertexLayerMaskAttribureId = this.addAttribute(
-                    GraphElementType.VERTEX,
-                    IntegerAttributeDescription.ATTRIBUTE_NAME,
-                    LAYER_MASK_ATTRIBUTE_LABEL,
-                    "Bitmask identifying the layers this vertex belongs to", 1, null);
-        }
-        if (transactionFilterBitmaskAttrId == Graph.NOT_FOUND) {
-            transactionFilterBitmaskAttrId = this.addAttribute(
-                    GraphElementType.TRANSACTION,
-                    IntegerAttributeDescription.ATTRIBUTE_NAME,
-                    LAYER_MASK_ATTRIBUTE_LABEL,
-                    "Bitmask identifying the layers this transaction belongs to", 1, null);
-        }
+        layerMaskSelectedAttributeId = LayersConcept.GraphAttribute.LAYER_MASK_SELECTED.get(this);
+        vertexLayerMaskAttribureId = LayersConcept.VertexAttribute.LAYER_MASK.get(this);
+        vertexLayerVisibilityAttributeId = LayersConcept.VertexAttribute.LAYER_VISIBILITY.get(this);
+        transactionLayerMaskAttrId = LayersConcept.TransactionAttribute.LAYER_MASK.get(this);
+        transactionLayerVisibilityAttributeId = LayersConcept.TransactionAttribute.LAYER_VISIBILITY.get(this);
+
+        // TODO: not sure about why we have to keep id's recorded
```

Comment:
The idea here was so that we didn't call to update the layer bit masks of Vertices or Transactions when an update to their Id was called. Every other instance that a Vertex or Transaction was altered will result in an update here. Just prevents a loop: update color -> trigger bitmask recalc -> change vertex bitmask -> trigger bitmask recalc. Value is set again and it continues to loop.

Reply:
Thanks @aldebaran30701. I updated the PR by removing the TODO comment but you saw it already. In that case a follow up question would be if all you want to do is see if the attribute is the vertex/transaction layer visibility attributes then rather than storing a copy of the attribute can we get it from the attribute description like:

```java
if (!LAYER_VISIBILITY_ATTRIBUTE_LABEL.equals(getAttributeDescription(attributeId))){
...
}
```

Reply:
That looks like it would work, if i remember correctly I check those Ids in ```updateLayerMask()``` somewhere, if that is the case would it be alright if we checked those the same way? Just wondering from a performance standpoint.

Reply:
>if that is the case would it be alright if we checked those the same way?

Yep we can. Though I'll do this is a separate PR as my focus is to get the unit tests over the line. It will also mean this PR is not making too many changes.

> Just wondering from a performance standpoint

Good point to keep in mind. I will do some profiling to compare both cases though I think it will be negligible because it will be looking up arrays, though the String compare will be the only issue.